jobsdb/jobsdb.go

```diff
@@ -2983,6 +2904,70 @@ func (jd *HandleT) StoreWithRetryEachInTx(ctx context.Context, tx StoreSafeTx, j
 	return res, err
 }
 
+func (jd *HandleT) internalStoreEachBatchRetryInTx(
+	ctx context.Context,
+	tx *Tx,
+	ds dataSetT,
+	jobBatches [][]*JobT) (
+	errorMessagesMap map[uuid.UUID]string, err error,
+) {
+	const (
+		savepointSql = "SAVEPOINT storeBatchWithRetryEach"
+		rollbackSql  = "ROLLBACK TO " + savepointSql
+	)
+
+	failAll := func(err error) map[uuid.UUID]string {
+		errorMessagesMap = make(map[uuid.UUID]string, len(jobBatches))
+		for i := range jobBatches {
+			errorMessagesMap[jobBatches[i][0].UUID] = err.Error()
+		}
+		return errorMessagesMap
+	}
+	defer jd.getTimerStat("store_jobs_retry_each_batch", nil).RecordDuration()()
+	_, err = tx.ExecContext(ctx, savepointSql)
+	if err != nil {
+		return failAll(err), nil
+	}
+	err = jd.internalStoreJobsInTx(ctx, tx, ds, lo.Flatten(jobBatches))
```

Comment:
one side-effect of using this here is that the `store_jobs` metric will be published additionally to `store_jobs_retry_each`

Reply:
Thankfully because the `table_prefix` is only `gateway` where this happens, we can excuse this(at least while inferring job storage times from the stats, but redundant stats would still exist).
We could pull out the stat part inside `internalStoreJobsInTx`, and have the caller publish said stat(maybe a wrapper?)